### DISCUSSION TITLE
URQL One-off Queries and Mutations + Cache Management Issue (Null Response)

### DISCUSSION THREAD

**Discussion by mbushpilot2b** [urql-graphql/urql#2783] [Q&A]
# Bug
With [One-off Queries and Mutations](https://formidable.com/open-source/urql/docs/basics/core/#one-off-queries-and-mutations) when I try to manage cache using `cache.invalidate()` in my `cacheExchange` I receive a `null` response back in my `.then(res)` of the `client.query()` or `client.mutation()`?.

## Expected
`cache.invalidate()` should not nullify the response to the `client.mutation().toPromise().then(res => console.log(res))`

### ⚠️ `cache.invalidate()` causes the `.then(res)` to return **`null`** every time ⚠️

# Code Example
## cacheExchange
```
const newClient = createClient({
  exchanges: [
    cacheExchange<GraphCacheConfig>({
      updates: {
        Mutation: {
          updateWorklist(result, _args, cache, _info) {
            console.log("updateWorklistCache", result) // result is defined here {updateWorklist: { id: 24, name: "test", visibility: "PRIVATE", filters: {…}, order: null } … }
            if (result?.updateWorklist?.id) {
              cache.invalidate({ __typename: 'Worklist', id: result?.updateWorklist?.id})
            }
          },
        },
      },
    }),
  ]
})
```

## client.mutation()
```
const updateWorklist = (id: string, worklist: any) =>
client
  .mutation(UpdateWorklistDocument, { id, worklist })
  .toPromise()
  .then((res) => {
    console.log('res', res) // returns {data: {updateWorklist: null}}
    message.success('Worklist saved')
    return {
      id: res.data.updateWorklist.id,
      name: res.data.updateWorklist.name,
      visibility: res.data.updateWorklist.visibility,
      filter: res.data.updateWorklist.filters?.filters,
      sort: res.data.updateWorklist.filters?.order,
    }
  })
  .catch((_err: any) => message.error('Error saving worklist'))
```

@kitten This is what I emailed you about today. Thank you so much for all you have done for the urql project 🚀

### Reproduction

https://codesandbox.io/s/goofy-feather-wtz7w1?file=/src/index.js:837-843

### Urql version

![Screen Shot 2022-11-03 at 8 33 07 PM](https://user-images.githubusercontent.com/4673917/199860188-73e7b540-7665-40dd-858f-060a80168d0c.png)

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/FormidableLabs/urql/discussions) should be used
- [X] Read the [docs](https://formidable.com/open-source/urql/docs/).
- [X] Follow our [Code of Conduct](https://github.com/FormidableLabs/urql/blob/main/CODE_OF_CONDUCT.md)

**Comment by mbushpilot2b** [comment]
Thank you @JoviDeCroock for your help in creating a reproduction of the issue!

In this case, we get the cache update for free without the cache invalidating because it's an update. But if I do a delete or create mutation I cannot invalidate my cache and I need to update my cache.

**Comment by mbushpilot2b** [comment]
@JoviDeCroock Let me know if there is anything I can do to help

**Comment by JoviDeCroock** [comment]
Hey,

I have analyzed this issue a bit this morning and came to the main cause of this issue, which resides [here](https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/src/cacheExchange.ts#L227-L242) so basically what happens is we write the result (this results in the automatic update as well as trigger the updater) and then we query the data again so we can hit all `resolvers` and do the appropriate data-transformations.

The catch here is that we can't switch these around else we risk writing transformed data _or_ reading stale data and we can't stop re-reading the data for mutations as else we stop using resolvers for data coming from mutations.

Putting this into this specific context, we set the data to null and when we re-read it we read the null rather than the data. Will think more about this next week as it's quite busy for me atm 😅

**Comment by mbushpilot2b** [comment]
@JoviDeCroock Hey, that makes sense, yes that's exactly what's happening. Because of this, I have to either force all queries to fetch every time or invalidate the cache and use the input to my mutation as the output, which is bad practice. So this is very good we found this bug. Seems like a little challenge to fix! Thank you for looking into this!

**Comment by mbushpilot2b** [comment]
@JoviDeCroock Btw one-off queries and mutations are very important for use with tools like XState where hooks are not appropriate. I appreciate your time and effort in replicating and thinking up a solution!

**Comment by JoviDeCroock** [comment]
No need to tag me on every message 😅 

Just to clarify this, this isn't really a show-stopper at all for most invalidations. The times you will use invalidate will be

- when creating entities (invalidate lists, ...)
- when deleting entities (invalidate lists, entity itself, ...)

Both these won't remove the response the update scenario however is an automatic update which makes it a bit harder to retain the response when you are using invalidate as the entity is actually still present but removed for some reason the consumer choose

**Reply by mbushpilot2b** [reply]
Thanks for the response!

> the update scenario however is an automatic update

Absolutely, however, an update can remove an item from a list that is filtered, or change its location in a list that is sorted. This logic is on server side, it is impossible for us to know if an update should invalidate a list or not, so we must invalidate it always on update mutation. If we don't items showing up in a random place in a list, or items showing that should be filtered out of the list. etc...

> Both these won't remove the response

Yeah without us manually invalidating. Which we cannot do, otherwise we get a `null` response, which is a bug.

> this isn't really a show-stopper at all for most invalidations

In our case it is for a majority of our mutations as the server logic filters/sorts our lists server side. Not all of our mutations but a lot of them.

**Comment by kitten** [comment]
Hiya, I converted this to a discussion.

As explained, when invalidation is used you're deleting data from the cache. When you delete the data that you just got from a response, then the cache read that is subsequently run doesn't have access to that data. Basically, even for a mutation, data is read according to how that mutation has updated the cache. This is important for how resolvers apply to certain data, and for consistency.

Generally, I'd regard this as an acceptable edge case. A cache miss can always occur on the mutation itself when said mutation deletes its own cache databin updaters. However, this certainly isn't what most people run into since it technically does not make much sense to delete data that a mutation has just provided and updated 😅

I'm not sure what you're trying to achieve in this case. If it's that you want all queries to just refetch, I'm doubting why you're trying to use a normalised cache. If you have a mutation and are trying to invalidate data in queries, you have to take care not to invalidate the entity of the mutation itself (which already updates according to the mutation response) but it's better to update or invalidate data that is indirectly affected

**Reply by mbushpilot2b** [reply]
Thanks for the response!

> Generally, I'd regard this as an acceptable edge case.

I would beg the differ, see below.

> If it's that you want all queries to just refetch, I'm doubting why you're trying to use a normalised cache.

Our project is large, cache management for some simple update/create/delete mutations can be handled easily by an updateQuery. However, in this case there is sorting/filtering applied on server side, so invalidation is our only option.

> However, this certainly isn't what most people run into since it technically does not make much sense to delete data that a mutation has just provided and updated 😅

In this case we use the returned deleted worklist id in a XState condition during transition to another state. Correct me if I am wrong, but it is best practice to use data returned by the mutation instead of the input into the mutation? 

### Comments on Create mutations
As for create entity mutations we want to invalidate a lists that may potentially contain that entity where complex server logic filters/sorts that list before returning the result to the client. So not being able to invalidate the entire list in this case leaves us with stale queries. Also we need the data returned by the create mutation to properly fill the context of XState, especially when the create mutation service has complex logic that may change the result of the mutation, say default values based on user context, time of day, etc...

**Reply by kitten** [reply]
I don't doubt that, but what I was saying is that you run invalidate your "Worklist" in an "updateWorklist" updater, hence, you're invalidating presumably what your mutation field returns. That's the edge case I'm referring to, because that's basically unsupported behaviour, hence, presumably that's what's causing your null response, as per what I can see in your reproduction.

The goal of a mutation updater is it to model how the cache is affected by mutations, so in other words how queries are affected by them. The invalidation is for edge cases to invalidate entire entities that mutations don't return or have access to and that you can't or don't reasonable want to update.

So, in other words, the issue stems from the updater deleting data that the mutation already has automatically updated, which subsequently invalidates data that the mutation contains. Hence, when the mutation updates its result against the cache, the cache dictates that, due to the invalidation, the result is partially encountering a cache miss.

In other words, there's no problem with updating links with cache.link (et al), there's no problem with invalidating items, you may also invalidate fields that your mutation does not contain, or manually fan out into other updates, or invalidate lists or update them as items are created or updated. That's all expected.
However, deleting the entity that a mutation just delivered from the cache will cause the mutation result to  be a cache miss.

Edit: If this is all clear btw, I also understand it's not feasible to write lots of small queries and fragments and to then use cache.updateQuery and cache.writeFragment for all of them. To be clear, this is very very rarely needed. The mutation will automatically update the cache with its data, for instance "updateWorklist" will update the "Worklist" automatically, and if this affects other queries, you likely only need to update relations to invalidate data outside the mutation's selection set.

**Reply by mbushpilot2b** [reply]
Thank you for your reply!

> That's the edge case I'm referring to, because that's basically unsupported behaviour

The reason this is not an edge case is because of server logic that is impossible to reproduce on the client. (Filtering/Sorting etc) In my company, half of our queries use server-side (sorting/filtering/pagination etc).

### Example
Say I have a `todo` I want to mutate `(delete/update/create)`, and I have a list called `todos` with `sorting, filtering, and/or pagination` applied on the **server side in my graphql service** DB query using knexjs. **I cannot replicate this logic on the client, nor would I want to.** So I cannot use `updateQuery("todos" etc...)`, I must invalidate the `todos` list. The only way to invalidate the `todos` list is to **invalidate a `todo` entity** in the `todos` list (as far as I am aware). So I do this, and I get a `null` response in the `client.mutation()` response.

I have to know what the id is of the `create` mutation especially, so I resort to using no invalidation, and instead, all components querying "todos" must use the `cache-and-network` request policy and re-render every time. This only works in my particular circumstance because I destroy the dropdown where the "todos" list appears on every close. Meaning it's fresh data when I open (and renders it, fetching new `todos` list. This makes the app slower and is 100% not ideal.

**Reply by kitten** [reply]
> The only way to invalidate the todos list is to invalidate a todo entity in the todos list (as far as I am aware

It's luckily (well, and by design 😅) not the only way. Instead you can also invalidate any other entity or field, so you could also invalidate the field that carries the list itself, which also pretty much conveys your intention of invalidating a specific list: https://formidable.com/open-source/urql/docs/api/graphcache/#invalidate

So, I'd suggest you go invalidate the specific field that the list is on

**Reply by mbushpilot2b** [reply]
So kitten thank you for the nudge in the right direction, I realized I could use:
```
  const key = 'Query';
  cache
    .inspectFields(key)
    .filter(field => field.fieldName === 'worklists')
    .forEach(field => {
      cache.invalidate(key, field.fieldKey);
    });
```

Found [here](https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#invalidating-individual-fields)

To invalidate the entire response of my query to fetch the list (not just the list itself, which in my case is useful!)

Now that I have a solution for invalidating lists on `update/create/delete` (`todos`), I only have one more problem, a solution for `todo` entity deletions:
* `update` to entity updates cache for free 🎉
* `create` entity updates cache for free 🎉
* **`delete` entity does not seem to remove from the cache for free, again, how can I delete say a `todo` entity by id from my cache without causing the `null` problem?**

The reason I ask is, these deleted `todo` entities would remain in my cache until the browser refreshes which is probably not ideal. Harmful? probably not, but yeah.  Any thoughts?

Btw thank you for the speedy responses, I really appreciate it!
